cmd/tikv-ctl/src/main.rs

```diff
@@ -665,9 +697,177 @@ fn compact_whole_cluster(
+            let start_key = start_key.clone();
+            let end_key = end_key.clone();
+            let cfg = cfg.clone();
+
+            load_regions_to_each_store(&pd_client, start_key, end_key)
+                .into_iter()
+                .for_each(|(store, leaders)| {
+                    let addr = pd_client.get_store(store).unwrap().address;
+                    leaders
+                        .into_iter()
+                        .filter(|r| region_ids.is_empty() || region_ids.contains(&r.get_id()))
+                        .for_each(|region| {
+                            // get key range from region.
+                            let mgr = Arc::clone(&mgr);
+                            let addr = addr.clone();
+                            let cfg = cfg.clone();
+
+                            let start_key = region.get_start_key().to_vec();
+                            let end_key = region.get_end_key().to_vec();
+                            let region_id = region.get_id();
+                            let prepare_wg = prepare_wg.clone();
+                            let work = prepare_wg.clone().work();
+                            tokio::spawn(async move {
+                                let debug_executor =
+                                    new_debug_executor(&cfg, None, Some(&addr), mgr);
+                                match debug_executor.flashback_to_version(
+                                    version,
+                                    region_id,
+                                    start_key,
+                                    end_key,
+                                    start_ts.into_inner(),
+                                    0,
+                                ) {
+                                    Ok(_) => drop(work),
+                                    Err(_) => prepare_wg.work_fail(),
+                                }
+                            });
+                        })
+                });
+
+            // Wait for finishing prepare flashback.
+            match tokio::time::timeout(Duration::from_secs(FLASHBACK_TIMEOUT), prepare_wg.wait())
+                .await
+            {
+                Ok(success) => {
+                    if success {
+                        break;
+                    } else {
+                        start_ts = pd_client.get_tso().await.unwrap();
+                        continue;
+                    }
+                }
+                Err(_) => {
+                    println!("prepare flashback timeout.");
+                    return;
+                }
+            }
+        }
+
+        // Start flashback for all regions.
+        let mut commit_ts = pd_client.get_tso().await.unwrap();
+        let flashback_wg = WaitGroup::new();
+        loop {
+            let start_key = start_key.clone();
+            let end_key = end_key.clone();
+            let cfg = cfg.clone();
+            load_regions_to_each_store(&pd_client, start_key, end_key)
```

Comment:
Can we make only one call during the prepare phase and then directly reuse the regions' information obtained earlier?